### ISSUE TITLE
AI generated book blurbs (and summaries)

### ISSUE THREAD

**Issue by bburns** [gutenbergtools/gutenbergsite#35]
I love Gutenberg books, but it's hard to know what a book is about without some searching around. 

It would be great if we could include a brief blurb with each book, and show them on the website. 

We could use ChatGPT or some other LLM - they were all trained using Gutenberg texts, so should know them already. 

eg "write a short blurb for frankenstein" - 

> In Mary Shelley's timeless classic Frankenstein, young scientist Victor Frankenstein defies nature by creating life from dead matter, only to be horrified by the monstrous being he brings into existence. Abandoned and shunned, the creature embarks on a quest for revenge against his creator, leading to a harrowing chase across Europe and into the icy Arctic. A profound exploration of ambition, isolation, and the consequences of playing God, Frankenstein remains a poignant and cautionary tale about the perils of unchecked scientific experimentation and the deep-seated human need for connection and empathy.

---

And you could vary the blurb length, etc. 

I asked ChatGPT for some Python code - 

---

First, install the `gutenbergpy` and `openai` libraries:

```sh
pip install gutenbergpy openai
```

Then, use the following script:

```python
import openai
import os
from gutenbergpy.textget import get_text_by_id
from gutenbergpy.gutenbergcache import GutenbergCache

# Set your OpenAI API key here
openai.api_key = os.getenv('OPENAI_API_KEY')

def generate_blurb(book_title, book_author):
    prompt = f"Generate a blurb for the book '{book_title}' by {book_author}."
    
    response = openai.Completion.create(
        engine="text-davinci-003",
        prompt=prompt,
        max_tokens=100
    )
    
    blurb = response.choices[0].text.strip()
    return blurb

def get_gutenberg_books(limit=10):
    cache = GutenbergCache.get_cache()
    metadata = cache.get_metadata()
    books = []
    
    for i, book_id in enumerate(metadata['id']):
        if i >= limit:
            break
        title = metadata['title'][i][0] if metadata['title'][i] else 'Unknown Title'
        author = metadata['author'][i][0] if metadata['author'][i] else 'Unknown Author'
        books.append({'title': title, 'author': author})
    
    return books

def main():
    books = get_gutenberg_books(limit=10)  # Fetch metadata for a limited number of books
    
    for book in books:
        title = book['title']
        author = book['author']
        blurb = generate_blurb(title, author)
        print(f"Blurb for '{title}' by {author}:\n{blurb}\n")

if __name__ == '__main__':
    main()
```

I haven't tested it yet though. 

Also, if you ask ChatGPT to summarize a book, it will say something like this, which could be nice to include too - 

---

**Title:** *Frankenstein; or, The Modern Prometheus*  
**Author:** Mary Shelley  
**Publication Date:** 1818

**Summary:**

*Frankenstein* is a novel about Victor Frankenstein, a young scientist who creates a grotesque creature in an unorthodox scientific experiment. The story is framed by letters from Captain Robert Walton to his sister, documenting Walton's own expedition to the North Pole, where he encounters Frankenstein.

### Plot Overview:

**Part I:**
- **Introduction and Letters:** Captain Robert Walton writes letters to his sister, detailing his journey to the North Pole and his encounter with Victor Frankenstein.
- **Victor's Background:** Victor Frankenstein recounts his early life in Geneva, his interest in science, and his studies at the University of Ingolstadt.
- **Creation of the Monster:** Victor becomes obsessed with creating life, eventually bringing to life a monstrous being from assembled body parts. Horrified by his creation, he abandons it.

**Part II:**
- **Aftermath and Consequences:** Victor falls ill from the shock of his creation. The monster, abandoned and alone, faces rejection and learns about human society through observation.
- **Monster's Story:** The creature confronts Victor and narrates his experiences. He describes his struggle for acceptance and his increasing hatred towards his creator for making him a pariah.
- **Creature's Demand:** The monster demands Victor create a female companion for him, promising to disappear from human society if his request is fulfilled.

**Part III:**
- **Creation of the Companion:** Victor reluctantly agrees but ultimately destroys the female creature before completion, fearing the potential consequences of creating another monster.
- **Revenge and Tragedy:** Enraged, the monster vows revenge, leading to the deaths of Victor’s loved ones, including his fiancée, Elizabeth, and his best friend, Clerval.
- **Pursuit and End:** Victor pursues the monster to the Arctic in an attempt to destroy his creation. He encounters Walton's expedition but dies from exhaustion. The monster appears, mourns Victor’s death, and vows to kill himself.

**Themes:**
- **Ambition and Hubris:** The dangers of overreaching ambition and the consequences of playing God.
- **Isolation and Loneliness:** The profound effects of isolation and the need for companionship.
- **Responsibility and Consequences:** The responsibilities of a creator towards their creation and the ethical implications of scientific experimentation.
- **Nature vs. Nurture:** The influence of environment and experiences on an individual's behavior and morality.

**Significance:**
*Frankenstein* is considered one of the earliest examples of science fiction and explores complex themes related to humanity, technology, and ethics. The novel raises enduring questions about the limits of scientific exploration and the moral responsibilities of creators.

---

I don't know the legalities involved with using their outputs though - need to look into that.

**Comment by eshellman** [comment]
Hi Brian!
About a year ago I tested ChatGPT and confirmed that it had NOT trained on PG texts. This year, their (very stupid) bots have been consuming PG texts, and I'm not sure whether their current model has been trained. You should be able to test by asking is obscure but distinctive queries about content in books that are NOT covered by Wikipedia, which ALL the AIs are training on. Frankenstein is really not representative of what an AI is likely to do.

It's not for me to say if PG wants to include "blurbs" in the website, but a couple of years ago I demonstrated how to do that with Wikidata. I was able to derive several thousand summaries that way. And all human generated.

I think it would be interesting to train a small model with PG full text and see how well it does on summarization and classification - I'm think of having some students work on this starting in the fall.

If you do more on this, I would certainly be interested in learning more, and we're always in need of talented contributors.

(I added an issue in the gutenbergpy repo alerting them of newer, cleaner txt and csv downloads, you don't really need to mess with it at all.) (ChatGPT likes 7 year old code)

**Comment by bburns** [comment]
Ah, thanks - I assumed they would use Gutenberg texts for training since it's all open. But maybe they lost too much information to be able to generate a blurb about more obscure books?

So I picked a random book (https://www.gutenberg.org/ebooks/46238) and asked ChatGPT for a blurb - 

---

Could you write a blurb for Mazes and Labyrinths by W H Matthews from 1922?

Certainly! Here is a blurb for "Mazes and Labyrinths" by W. H. Matthews, published in 1922:

Mazes and Labyrinths by W. H. Matthews is a comprehensive exploration of the fascinating world of labyrinthine structures, tracing their history from ancient times to the early 20th century. Published in 1922, this seminal work delves into the architectural, cultural, and symbolic significance of mazes and labyrinths across various civilizations. Matthews combines meticulous research with engaging storytelling, offering readers a detailed account of these enigmatic designs found in classical mythology, medieval cathedrals, and elaborate garden landscapes. Richly illustrated and thoroughly documented, this book remains an essential reference for anyone interested in the enduring mystery and allure of mazes and labyrinths.

---

Claude.ai didn't do as well - 

I can certainly try to write a blurb for "Mazes and Labyrinths" by W. H. Matthews from 1922. However, I should note that while I have general knowledge about mazes and labyrinths, I don't have specific information about this particular book's contents. I'll create a plausible blurb based on what such a book from that era might contain, but please be aware that this is a speculative description...

---

This says the models all use Gutenberg though - https://our-hometown.com/overview-of-data-used-to-train-language-models/

So, I'll try out the code, and look into the legality...

**Comment by jhellingman** [comment]
Nice to see the blurbs for fiction books, but they have numerous factual mistakes (e.g., claiming that a book was written in the early 21st century, etc.) They also have a typical blurb-like appearance to entice people to read. No problem with that per se, but the guesses the LLM makes are often a bit too wild.

I think a better approach would be not to rely on the training data of the LLM, but just supply the entire text in the context, and then have it summarized. (And if the book is too long, summarize things chapter-by-chapter, then feed those chapter summaries for a full book summary).

Finally, for books not in English, adding a summary in their own language would be nice as well.

**Comment by bburns** [comment]
Yeah, I had wanted blurbs so I could be enticed into reading some good old books, without spoilers! A more complete description or summary could be rendered as well - then a website could show whichever the user wanted. 

My initial thought was the same, to feed it a whole book to summarize, but wasn't sure how well it would work with the context limitations - though breaking things up by chapter like you say might work. But with 60k books, it could get expensive (?). 

And I didn't think to fact check the blurb - hallucinations aren't good! Feeding it a book at a time might cut that down - it would be interesting to compare the output and see if it was more accurate. 

Anyway, I'd love to work on this, but am swamped - anyone else is welcome to continue with it!

**Comment by jhellingman** [comment]
I think for fiction, an enticing blurb is good enough; for non-fiction, a summary is better (and we could pull that from Wikipedia as well).

With regard to the cost: why not use our user's free accounts on the various services: when they open a catalogue page without a blurb or summary, just add a button to add it, and client side code will push it to the service and retrieve the details.

Very interesting stuff, but I am also swamped with PGDP work (besides all the other things in RL).

**Comment by gbnewby** [comment]
You can find an announcement of the 72000+ summaries that are now part of
the PG metadata, here:

***@***.***/thread/MTHHI3TD7YXD3EHLKVBBA57KRBBWRI72/

We do not have capability to confirm quality of non-English blurbs, so they
are only in English.

The intention is to do a monthly update to any new books, and to adjust the
prompt if needed. When technology improves significantly, we can
re-generate all the blurbs.

**Comment by eshellman** [comment]
I note that in loading the automated "blurbs", all the human-generated summaries in the 520 attribute were wiped.

**Comment by eshellman** [comment]
I am mistaken, the blurbs are added as additional summaries and the old ones are still there

**Comment by bburns** [comment]
Oh, that's great - can you fix this link?

> You can find an announcement of the 72000+ summaries that are now part of
> the PG metadata, here:
>  ***@***.***/thread/MTHHI3TD7YXD3EHLKVBBA57KRBBWRI72/

**Comment by gbnewby** [comment]
github stripped the link, sorry. Just visit https://lists.pglaf.org and see the gmonthly archives - it's the most recent message (October 2024)

**Comment by bburns** [comment]
@gbnewby Thanks, here's the text -

--- 

AI Roles for Free eBooks

Have you noticed that most books now have summaries and browsing categories?

Project Gutenberg has been corresponding with innovators seeking to leverage today's new artificial intelligence (AI) systems to improve access to free literature, as well as increase human capabilities for benefitting from literature.

In 2022-2023, we partnered with Microsoft and others to make thousands of new audiobooks, leveraging AI techniques to improve the quality of the audio speech. You can find these via a section on the main page at https://www.gutenberg.org/.

In the September 2024 newsletter, we celebrated the addition of AI-derived browsing categories in the Bookshelves section of the website, making it easier to browse the collection. You can find those categories here: https://www.gutenberg.org/ebooks/bookshelf/

We then worked with the same programmers to provide automated summaries of nearly every book in the collection. You can find those summaries on book landing pages. These summaries are intended to be helpful for people trying to decide what book to read, or to get an idea of what a book is about.

If you spot errors in summaries, let us know. Summaries of most books are based only on the first 12,000 characters, because the costs would have been too high for if we included all of every book.

We have also been corresponding with another programmer seeking to instruct AI technologies to "read" books from Project Gutenberg, summarize them, and answer questions about them. We hope this might be described further in a future newsletter.

---

12k characters is about 2k words, or 8 pages. 

Eg here's a nonfiction summary - https://www.gutenberg.org/ebooks/65903

> "Psychology of the Unconscious" by C. G. Jung is a scientific publication written in the early 20th century. This work is a study that explores the transformations and symbolisms of human libido, contributing to our understanding of the evolution of thought and psychological conflicts throughout history. It delves into the connections between ancient mythologies and modern psychological concepts, particularly focusing on the unconscious mind's influence on human behavior and relationships. The beginning of "Psychology of the Unconscious" introduces the broad theme of psychological exploration and the significance of understanding the unconscious. Jung emphasizes the longstanding connections between individual psychology and the historical psyche, using the example of the Oedipus legend to illustrate the continuity of human conflicts across time. The opening chapters articulate the importance of dreams as symbolic expressions of latent desires, setting the stage for deeper investigations into the workings of the psyche and the role of the libido in shaping human experiences and interpersonal dynamics. (This is an automatically generated summary.)

And here's a fiction - https://www.gutenberg.org/ebooks/30255

> "The Skull" by Philip K. Dick is a science fiction novella written in the early 1950s. The story delves into the complexities of time travel and the philosophical implications of altering the past, focusing on the themes of identity, mortality, and the consequences of one's actions. It presents a speculative narrative in which a man named Conger is hired to assassinate a figure known as the Founder, who has been dead for centuries, using nothing but the deceased's skull as a means of identification. In the novella, Omar Conger, a hunter and a rogue, is approached while in prison by a council speaker who offers him a chance at redemption: he must travel back in time to eliminate the Founder before he can influence society with his doctrine of non-violence. Conger is equipped with a futuristic weapon and the skull of the Founder, which is purported to be the only way he can identify his target. As he navigates the past, Conger encounters various townspeople and begins to unravel the implications of his mission, grappling with existential questions about fate and purpose. The story culminates in a paradoxical acceptance of his fate, showcasing Dick's talent for weaving complex philosophical dilemmas within engaging narratives. (This is an automatically generated summary.)

I deleted a sentence with spoilers. 

Anyway, this is great - it really opens up the world of these books so much more!

**Comment by mrcook** [comment]
> I deleted a sentence with spoilers. 

And therein lies the problem. Each and every summary would need to be checked by a human who has actually read the book. No?

I assume the generated blurbs are only being shown on gutenberg.org as a fallback when there's no existing text?

**Comment by bburns** [comment]
Yeah, a blurb is generally something without spoilers - I'd be interested to see what prompt they used to make those.

Sometimes you'd just want to see blurbs, sometimes full summaries...

**Comment by eshellman** [comment]
further development of the AI summaries should go in a new issue, or perhaps a new repo. https://github.com/gutenbergtools/website_update would be a good place to organize
